Deepseek V3.1 native tool calling support (OpenAI Style)

createthis opened the pull request:
This PR enables DeepSeek V3.1 thinking mode as the default. Disable with `--reasoning-budget 0`.

It also implements tool calling support.

Addresses https://github.com/ggml-org/llama.cpp/issues/15496

My understanding is that this is a continuation of https://github.com/ggml-org/llama.cpp/pull/9639 for DeepSeek V3.1 specifically.

openhands-agent committed:
Add DeepSeek V3.1 thinking mode support

- Added COMMON_CHAT_FORMAT_DEEPSEEK_V3_1 enum value
- Created common_chat_params_init_deepseek_v3_1() function (currently uses R1 implementation)
- Created common_chat_parse_deepseek_v3_1() function that handles V3.1 thinking format:
  - Extracts reasoning content before '</think>' tag into reasoning_content
  - Extracts regular content after '</think>' tag into content
  - No opening '<think>' tag in V3.1 format
- Added detection logic for V3.1 templates based on pattern: 'message['prefix'] is defined and message['prefix'] and thinking'
- Added V3.1 case to parsing switch statement

This addresses the issue where V3.1 outputs reasoning content followed by '</think>' and then regular content without the opening '<think>' tag.

createthis committed:
Another attempt by V3.1 non-thinking

createthis committed:
Fix test, but it's not asserting anything.

createthis committed:
Ignore vim swap files in tests dir

createthis committed:
Update the test

createthis committed:
Try using try_find_literal instead of regex

createthis committed:
passing test

createthis committed:
Revert "Try using try_find_literal instead of regex"

This reverts commit c50d887ec2780dd9e6b8b397e92347d3db8d5575.

createthis committed:
Remove unnecessary change

createthis committed:
Remove comment

createthis committed:
Add code to handle non-thinking mode.

createthis committed:
Try to set message['prefix'] when thinking is enabled.

createthis committed:
This fixes reasoning, but breaks normal content. We need state in the
chat parser.

createthis committed:
DeepSeek V3.1 thinking is now the default. Disable with `--reasoning-budget 0`.

createthis committed:
Simplify (DeepSeek V3.1 reasoning)

CISC commented on common/chat.cpp:
```diff
@@ -1312,6 +1313,27 @@ static common_chat_params common_chat_params_init_deepseek_r1(const common_chat_
+        data.thinking_forced_open = true;
+    }
+    return data;
+}
```
I realize you probably didn't want to tackle tool calls in this PR, but I'm pretty sure this will break when there are tool calls if you don't.

createthis replied:
@CISC I had ChatGPT 5 generate a python script for a simple tool call test: https://github.com/createthis/llama_cpp_deepseek_v3_1_think_tags/blob/main/tool_test.py

This works for me with `gpt-oss-120b` and `master`. We can use this as our control test.

I recorded the successful conversations with `tool_test.py` and `gpt-oss-120b` twice, once without `--verbose` set in `llama.cpp` and once with. The files for the recordings are in the same repo:
- full_traffic_master_tool_test_gpt_oss_120b.mitm
- full_traffic_master_tool_test_verbose_gpt_oss_120b.mitm

createthis replied:
@CISC I ran `tool_test.py` against `master` running `DeepSeek-V3.1` using this start command:

```bash
./build/bin/llama-server \
    --model /data2/DeepSeek-V3.1-GGUF/UD-Q4_K_XL/DeepSeek-V3.1-UD-Q4_K_XL-00001-of-00008.gguf \
    --alias DeepSeek-V3.1:671b-q4_k_xl \
    --no-webui \
    --numa numactl \
    --threads 32 \
    --ctx-size 163840 \
    --n-gpu-layers 62 \
    -ot "blk\.(3|4|5|6|7|8|9|10|11)\.ffn_.*=CUDA0" \
    -ot exps=CPU \
    -ub 4096 -b 4096 \
    --seed 3407 \
    --temp 0.6 \
    --top-p 0.95 \
    --min-p 0.1 \
    --log-colors \
    --flash-attn \
    --host 0.0.0.0 \
    --verbose \
    --jinja \
    --prio 2 \
    --port 11434
```

The test failed. The transcript is in the above linked repo as file `full_traffic_master_tool_test_verbose_deepseek_v3_1.mitm`.

Next, I'll try this branch, then I'll try this branch with your suggested changes.

createthis replied:
Also, I'm pretty confident I can fix R1's tool calling support too at this point.

CISC commented on tests/.gitignore:
```diff
@@ -1,4 +1,5 @@
 *
 !*.*
 *.o
+*.swp
```
```suggestion
```

createthis replied:
Why remove this? It already exists in the `.gitignore` in the directory above: https://github.com/ggml-org/llama.cpp/blob/b730706a49e576fb882dc34d9966345778b3ab0b/.gitignore#L21

CISC replied:
Merely because it's a totally unrelated change, if you need them ignored make a separate PR.

createthis replied:
ok, removed

CISC commented on tests/test-chat-parser.cpp:
```diff
@@ -186,6 +186,24 @@ static void test(const std::string & input, bool is_partial, const std::vector<s
   assert_equals(expected, args_paths.size() == 1 && args_paths[0].empty() ? js->value.get<std::string>() : js->value.dump());
 }
+
+static void test_deepseek_v3_1() {
```
This should be within `test_reasoning` instead.

createthis replied:
This was moved btw

createthis replied:
Ah, I see what happened. I forgot to commit it. Should be changed now.

createthis committed:
Fix sign inversion bug

createthis committed:
Add some tool calling code (not working).

createthis committed:
Tool calls working in non-reasoning mode.

createthis committed:
Attempt a unit test for tool call parsing.

createthis committed:
Passing test

createthis committed:
Add tests for both happy path and broken fenced DeepSeek V3.1 tool call variants.

createthis committed:
Passing DeepSeek V3.1 tool call tests, but model is not working.

createthis committed:
Revert assistance response prefill change. Not my monkeys.

createthis committed:
Add fenced_thinking unit test variant. Passes, but thinking tool calling
still isn't working for some reason.

createthis committed:
Tests pass in reasoning mode. Also e2e tool test passes.

createthis committed:
Make a copy of the parse_json_tool_calls function for deepseek-v3.1 so
as to not accidentally introduce regressions.

createthis committed:
Merge pull request #7 from createthis/deepseek_3_1_thinking_mode_danger

Tool calls work in thinking and non-thinking modes. However, I've introduced a regression in streaming mode where reasoning content initially comes through as regular content. I need to think about how to deal with this long term.

createthis committed:
Fix thinking_forced_open logic. tool calling broken. Need to add another
test case.

createthis committed:
That's what I get for cargo culting a newline.

createthis committed:
Add multi tool call test for deepseek v3.1 non-reasoning

createthis committed:
Merge branch 'master' into deepseek_3_1_thinking_mode

createthis committed:
Move test, remove .gitignore change

createthis committed:
Place deepseek-v3.1 reasoning test directly into existing reasoning
function per CISC's request.

createthis committed:
Address whitespace CI failure.

createthis commented:
> Also, address this CI failure: https://github.com/ggml-org/llama.cpp/actions/runs/17274346584/job/49214116367

@CISC you got it [9056707](https://github.com/ggml-org/llama.cpp/pull/15533/commits/905670711e64c8d65c3b8ec1e8842b9164f4764d)

CISC commented on tests/test-chat-parser.cpp:
```diff
@@ -186,6 +200,141 @@ static void test(const std::string & input, bool is_partial, const std::vector<s
+    }
+}
+
+static void test_deepseek_v3_1_tool_calls() {
```
These tests are fine, but it would be even better to add tests in `test-chat.cpp` too where you have `assert_msg_equals`.

createthis replied:
@CISC You got it: [ec984da](https://github.com/ggml-org/llama.cpp/pull/15533/commits/ec984da15252a7527cf47b886eba315d3eda2ae9)

createthis committed:
Merge two assert_equals per CISC's request.

createthis commented on tests/test-chat-parser.cpp:
```diff
@@ -186,6 +200,141 @@ static void test(const std::string & input, bool is_partial, const std::vector<s
+        std::ostringstream oss; oss << label << "\nExpected: " << expected << "\nActual: " << actual;
+        throw std::runtime_error(oss.str());
+    }
+}
```
You got it [a406d6a](https://github.com/ggml-org/llama.cpp/pull/15533/commits/a406d6a2d17db32f69b9105ded88f471180e8622)

createthis committed:
Add DeepSeek-V3.1 tests to tests/test-chat.cpp per CISC's request.

createthis commented on common/chat.cpp:
```diff
@@ -665,6 +666,93 @@ static std::string wrap_code_as_arguments(common_chat_msg_parser & builder, cons
+ * Takes a prefix regex that must have 1 group to capture the function name, a closing suffix, and expects json parameters in between.
+ * Aggregates the prefix, suffix and in-between text into the content.
+ */
+static void parse_json_tool_calls_deepseek_v3_1(
```
Now I know if I replace it, it breaks Functionary v3.2 tests at a minimum.

createthis replied:
I merged them using an optional `update_cursor` argument.

createthis replied:
@sgoll It works in the builtin webui with `--reasoning-budget 0`. That's something at least. Still investigating.

createthis replied:
@sgoll The builtin webui uses `reasoning_format: none`, but I wasn't supporting it. This should be fixed by [7795594](https://github.com/ggml-org/llama.cpp/pull/15533/commits/77955943d7bf54fbad1436191f498795186d57ce).

createthis replied:
<img width="1142" height="709" alt="Screenshot 2025-08-31 at 10 44 26 PM" src="https://github.com/user-attachments/assets/5503a304-1f6f-478d-9ba8-ade19059eeb6" />


Note that `reasoning_format: none` is pretty basic. Open WebUI formats it better because it uses `reasoning_format: auto`:

<img width="1131" height="462" alt="Screenshot 2025-08-31 at 10 45 32 PM" src="https://github.com/user-attachments/assets/f7f484a0-32ee-4b2f-8e64-c2503f4d828b" />

createthis committed:
Merge branch 'master' into deepseek_3_1_thinking_mode

createthis committed:
Merge deepseek V3.1 and regular parse_json_tool_calls() function
behaviors by adding optional update_cursor argument.

CISC commented on common/chat.cpp:
```diff
@@ -678,7 +679,8 @@ static void parse_json_tool_calls(
     bool allow_raw_python = false,
-    const std::function<std::string(const common_chat_msg_parser::find_regex_result & fres)> & get_function_name = nullptr) {
+    const std::function<std::string(const common_chat_msg_parser::find_regex_result & fres)> & get_function_name = nullptr,
+    bool update_cursor = false) {
```
I'm struggling to understand why the `update_cursor` changes are necessary, why exactly do you need to do this?

I see that the tests fail without it, but the tests are also altered compared to all the other formats...

createthis replied:
> I see that the tests fail without it

@CISC Correct. The regexes simply do not work without this change.

createthis replied:
@CISC What's incorrect about the test?

createthis replied:
@CISC I'm not super smart. Dumb it down for me. Are you asking for a change? Is a specific line incorrect? Give me something to go on here.

createthis replied:
@CISC FYI, I still can't get the tests to pass with the new grammar from [a7316b0](https://github.com/ggml-org/llama.cpp/pull/15533/commits/a7316b0696aea460db6465c26161c2fd95bb25fb) unless I set `update_cursor` to `true`.

If you really need me to give you a presentation on why this is the case, I can build a step-by-step flow/state output for you, but it's going to be a fair bit of work.

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC commented on common/chat.cpp:
````diff
@@ -1313,6 +1334,71 @@ static common_chat_params common_chat_params_init_deepseek_r1(const common_chat_
+                std::string name = function.at("name");
+                auto parameters = function.at("parameters");
+                builder.resolve_refs(parameters);
+                tool_rules.push_back(builder.add_rule(name + "-call",
+                    "( \"<｜tool▁call▁begin｜>\" )? \"function<｜tool▁sep｜>" + name + "\\n"
+                    "```json\\n\" " + builder.add_schema(name + "-args", parameters) + " "
+                    "\"```<｜tool▁call▁end｜>\""));
````
This does not match `function_regex`.

createthis replied:
This code was cargo culted from the R1 code. I see the tests fail with it removed. I'll investigate. I have to cook some swordfish for my daughter first.

createthis replied:
I logged this out:

```cpp
std::string rule_name = name + "-call";
std::string rule_pattern = "( \"<｜tool▁call▁begin｜>\" )? \"function<｜tool▁sep｜>" + name + "\\n"
                           "```json\\n\" " + builder.add_schema(name + "-args", parameters) + " "
                           "\"```<｜tool▁call▁end｜>\"";
LOG_DBG("%s: add_rule: \nrule_name: %s\nrule_pattern: %s\n", __func__, rule_name.c_str(), rule_pattern.c_str());
```

## This logs

operator(): add_rule: 
rule_name: `special_function-call`
rule_pattern: `( "<｜tool▁call▁begin｜>" )? "function<｜tool▁sep｜>special_function\n```json\n" special-function-args "```<｜tool▁call▁end｜>"`

## function_regex
function_regex is `(?:<｜tool▁call▁begin｜>)?function<｜tool▁sep｜>([^\n]+)\n```json\n`

## testing function_regex
Here's a link to a regex tester that implements this pattern: https://regex101.com/r/WUCUga/1
And a screenshot of the tester for good measure:
<img width="1226" height="439" alt="Screenshot 2025-08-31 at 7 53 33 PM" src="https://github.com/user-attachments/assets/16341120-de23-41fc-a321-3bf0108b084e" />

## ???
Admittedly, I don't fully understand llama.cpp's grammar subsystem, but this looks like it works to me. Also, if I remove this code the tests fail. Help me out. What are we talking about here?

CISC replied:
It fails on this test:
https://github.com/ggml-org/llama.cpp/blob/77955943d7bf54fbad1436191f498795186d57ce/tests/test-chat.cpp#L1796-L1798

createthis replied:
@CISC No, it fails on the first `test_template` call, which was cargo culted from the R1 tests:

https://github.com/ggml-org/llama.cpp/blob/77955943d7bf54fbad1436191f498795186d57ce/tests/test-chat.cpp#L1327

## changes to repro

```diff
diff --git a/common/chat.cpp b/common/chat.cpp
index 1236e766..113eec90 100644
--- a/common/chat.cpp
+++ b/common/chat.cpp
@@ -1365,10 +1365,6 @@ static common_chat_params common_chat_params_init_deepseek_v3_1(const common_cha
                 std::string name = function.at("name");
                 auto parameters = function.at("parameters");
                 builder.resolve_refs(parameters);
-                tool_rules.push_back(builder.add_rule(name + "-call",
-                    "( \"<｜tool▁call▁begin｜>\" )? \"function<｜tool▁sep｜>" + name + "\\n"
-                    "```json\\n\" " + builder.add_schema(name + "-args", parameters) + " "
-                    "\"```<｜tool▁call▁end｜>\""));
             });
             // Distill Qwen 7B & 32B models seem confused re/ syntax of their tool call opening tag,
             // so we accept common variants (then it's all constrained)
diff --git a/tests/test-chat.cpp b/tests/test-chat.cpp
index 4bcbf97a..42d79508 100644
--- a/tests/test-chat.cpp
+++ b/tests/test-chat.cpp
@@ -1755,6 +1755,7 @@ static void test_template_output_parsers() {
                 /* is_partial= */ false,
                 {COMMON_CHAT_FORMAT_SEED_OSS}));
     }
+    common_log_set_verbosity_thold(LOG_DEFAULT_DEBUG);
     {
         auto tmpls = read_templates("models/templates/deepseek-ai-DeepSeek-V3.1.jinja");
         std::vector<std::string>   end_tokens{ "<｜end▁of▁sentence｜>" };
@@ -1765,8 +1766,11 @@ static void test_template_output_parsers() {
             assert_equals(true, params.thinking_forced_open);
         }
 
+        LOG_DBG("%s: here0\n", __func__);
         test_templates(tmpls.get(), end_tokens, message_assist, tools, "</think>Hello, world!\nWhat's up?", /* expect_grammar_triggered= */ false);
+        LOG_DBG("%s: here0.1\n", __func__);
         test_templates(tmpls.get(), end_tokens, message_assist_thoughts, tools, "</think>Hello, world!\nWhat's up?", /* expect_grammar_triggered= */ false);
+        LOG_DBG("%s: here0.2\n", __func__);
         assert_msg_equals(
             simple_assist_msg("Hello, world!\nWhat's up?", "I'm\nthinking"),
             common_chat_parse(
```

## result
Note the `here0` but no `here0.1`:

```
# Reading: models/templates/deepseek-ai-DeepSeek-V3.1.jinja
Template supports tool calls but does not natively describe tools. The fallback behaviour used may produce bad results, inspect prompt w/ --verbose & consider overriding the template.
test_template_output_parsers: here0
Template supports tool calls but does not natively describe tools. The fallback behaviour used may produce bad results, inspect prompt w/ --verbose & consider overriding the template.
Template supports tool calls but does not natively describe tools. The fallback behaviour used may produce bad results, inspect prompt w/ --verbose & consider overriding the template.
unsupported grammar, left recursion detected for nonterminal at index 3/home/jesse/llama.cpp/build-ci-debug/bin/libggml-base.so(+0x51511)[0x7a96c41c2511]
```

## analysis

I don't know why `tools` is using python:

https://github.com/ggml-org/llama.cpp/blob/77955943d7bf54fbad1436191f498795186d57ce/tests/test-chat.cpp#L213

I'm just trying to get V3.1 working, not fix the broken R1 code.

createthis replied:
@CISC I'm investigating this further by diving into the llama.cpp EBNF grammar subsystem.

createthis replied:
@CISC I was thinking about this on the way home from class today. When `function` was in the grammar it was totally outputting `function`. When I took it out, it stopped. It must be reading the grammar. Does the grammar get sent to the model as part of the request?

createthis replied:
Ok, I reverted to [3dee75a](https://github.com/ggml-org/llama.cpp/pull/15533/commits/3dee75a1a236ca7fcd53a2ef53d66ffc8180a9c7) locally and couldn't reproduce the outputting of `function` in the verbose output. Maybe I was hallucinating that behavior.

createthis committed:
DeepSeek V3.1 fix reasoning_format none

createthis committed:
Merge branch 'master' into deepseek_3_1_thinking_mode

createthis committed:
Strip grammar down to strictly what we expect based on model card. Throw
out parts we cargo culted from R1 that don't make sense.

createthis committed:
Merge pull request #8 from createthis/deepseek_3_1_thinking_mode_experimental

Strip grammar down to strictly what we expect based on model card. Throw

createthis committed:
Update tests/test-chat-parser.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Merge branch 'master' into deepseek_3_1_thinking_mode

createthis committed:
DeepSeek V3.1 - Add edge case where thinking is forced open, there is
tool calling in the reasoning content, but then the model just stops the
output without closing the </think> tag, so it's not a partial. In this
case, use the tool call in the reasoning content.

createthis commented:
@CISC @ggerganov Let me know if you want any more changes, otherwise please merge. This is working well on my end.

createthis committed:
Merge branch 'master' into deepseek_3_1_thinking_mode

createthis committed:
DeepSeek V3.1 - simplify update_cursor

createthis commented:
@CISC @ggerganov I simplified `update_cursor` with [26b02fa](https://github.com/ggml-org/llama.cpp/pull/15533/commits/26b02fa8ac248bf47f7c550f1d015fb9081a5499). I also performed an extensive analysis of the cause. You can view my notes here: https://gist.github.com/createthis/dc3098c3abb4ff809d0291c91322f512

CISC commented on common/chat.cpp:
```diff
@@ -692,17 +693,20 @@ static void parse_json_tool_calls(
     const std::optional<common_regex> & block_close,
     bool allow_raw_python = false,
-    const std::function<std::string(const common_chat_msg_parser::find_regex_result & fres)> & get_function_name = nullptr) {
+    const std::function<std::string(const common_chat_msg_parser::find_regex_result & fres)> & get_function_name = nullptr,
+    bool update_cursor = false) {
```
```suggestion
    const std::function<std::string(const common_chat_msg_parser::find_regex_result & fres)> & get_function_name = nullptr) {
```

createthis committed:
Update common/chat.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update common/chat.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Update common/chat.cpp

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

createthis committed:
Fix indent
